Failed tests (11584): TestSystemConfigGossip

The following test appears to have failed:

[#11584](https://circleci.com/gh/cockroachdb/cockroach/11584):

```
I0119 12:19:25.439421 742 kv/range_cache.go:189  adding descriptor: key=/Meta2/"\x93" desc=range_id:1 start_key:"" end_key:"\223" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
W0119 12:19:25.440865 742 rpc/send.go:198  Node.Batch: error reply: server/node.go:494: node 1 stopped
I0119 12:19:25.441372 742 client/db.go:485  failed batch: failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
E0119 12:19:25.441766 742 storage/queue.go:123  [split] range=3 [/Table/12-/Max): error: storage/split_queue.go:100: unable to split range=3 [/Table/12-/Max) at key "/Table/13/0": failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
I0119 12:19:25.444107 742 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestSystemConfigGossip (1.06s)
	server_test.go:566: mismatch: expected name:"foo" id:2 , got {Name:foo ID:1 Privileges:<nil>}
=== RUN   TestStatusLocalStacks
I0119 12:19:25.508942 742 rpc/clock_offset.go:153  monitoring cluster offset
I0119 12:19:25.537954 742 raft/raft.go:446  [group 1] 1 became follower at term 5
I0119 12:19:25.538668 742 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0119 12:19:25.539031 742 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0119 12:19:25.539372 742 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0119 12:19:25.539547 742 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0119 12:19:25.539905 742 raft/raft.go:487  [group 1] 1 became leader at term 6
I0119 12:19:25.543128 742 storage/replica_command.go:1214  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1453205966.542s
--
I0119 12:19:35.114486 742 client/db.go:485  failed batch: failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
W0119 12:19:35.114745 742 ts/db.go:96  error writing time series data: failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
I0119 12:19:35.116841 742 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestMetricsRecording (1.85s)
	<autogenerated>:28: prior leak detected, leaktest disabled
FAIL
FAIL	github.com/cockroachdb/cockroach/server	31.783s
=== RUN   TestNodeStatusMonitor
--- PASS: TestNodeStatusMonitor (0.01s)
=== RUN   TestNodeStatusRecorder
--- PASS: TestNodeStatusRecorder (0.03s)
=== RUN   TestRuntimeStatRecorder
I0119 12:19:25.007588 773 server/status/runtime.go:137  runtime stats: 11 goroutines, 3.04MiB active, 50000000.00cgo/sec, 544530.00/388950.00 %(u/s)time, 9051.40 %gc (2x)
--- PASS: TestRuntimeStatRecorder (0.01s)
=== RUN   TestNodeEventFeed
--- PASS: TestNodeEventFeed (0.00s)
=== RUN   TestServerNodeEventFeed

```
Please assign, take a look and update the issue accordingly.

--- a/gossip/infostore.go
+++ b/gossip/infostore.go
@@ -205,19 +205,20 @@ func (is *infoStore) getNodes() map[int32]*Node {
 // invoked whenever new info for a gossip key matching pattern is
 // received. The callback method is invoked with the info key which
 // matched pattern. Returns a function to unregister the callback.
+// Note: the callback may fire immediately after being unregistered.
 func (is *infoStore) registerCallback(pattern string, method Callback) func() {
 	re := regexp.MustCompile(pattern)
 	cb := &callback{pattern: re, method: method}
 	is.callbacks = append(is.callbacks, cb)
 	if err := is.visitInfos(func(key string, i *Info) error {
 		if re.MatchString(key) {
-			// Run callbacks in a goroutine to avoid mutex reentry.
-			go method(key, i.Value)
+			is.runCallbacks(key, i.Value, method)
 		}
 		return nil
 	}); err != nil {
 		panic(err)
 	}
+
 	return func() {
 		for i, targetCB := range is.callbacks {
 			if targetCB == cb {
@@ -240,10 +241,13 @@ func (is *infoStore) processCallbacks(key string, content roachpb.Value) {
 			matches = append(matches, cb.method)
 		}
 	}
+	is.runCallbacks(key, content, matches...)
+}
 
-	// Add the matching callbacks to the callback work list.
+func (is *infoStore) runCallbacks(key string, content roachpb.Value, callbacks ...Callback) {
+	// Add the callbacks to the callback work list.
 	f := func() {
-		for _, method := range matches {
+		for _, method := range callbacks {
 			method(key, content)
 		}
 	}
@@ -255,8 +259,7 @@ func (is *infoStore) processCallbacks(key string, content roachpb.Value) {
 	// callbacks are run in order such that if a key is updated twice in
 	// succession, the second callback will never be run before the first.
 	go func() {
-		// Grab the callback mutex to prevent another goroutine from invoking
-		// callbacks.
+		// Grab the callback mutex to serialize execution of the callbacks.
 		is.callbackMu.Lock()
 		defer is.callbackMu.Unlock()
 

